build: bump lightning-onion to fat errors

joostjager opened the pull request.

hieblmi commented on lnwire/features.go:
```diff
@@ -141,6 +141,14 @@ const (
 	// addresses for cooperative closure addresses.
 	ShutdownAnySegwitOptional FeatureBit = 27
 
+	// FatErrorsRequired is an required feature bit that signals
```
nit: an -> a

joostjager commented on htlcswitch/hop/error_encryptor.go:
```diff
@@ -79,22 +96,80 @@ type ErrorEncrypter interface {
+
+func (s *SphinxErrorEncrypter) UpgradeToFatErrors() error {
+	encryptor, err := sphinx.NewOnionFatErrorEncrypter(
+		s.router, s.EphemeralKey, FatErrorStructure,
```
It may be better to not use a hard-coded structure here and instead let the sender specify the structure via the onion (https://github.com/lightning/bolts/pull/1044#discussion_r1069623771).

joostjager replied:
Implemented this.

joostjager commented on htlcswitch/hop/error_encryptor.go:
```diff
@@ -79,22 +96,80 @@ type ErrorEncrypter interface {
+		holdTimeMs := s.getHoldTimeMs()
+
+		var payload [8]byte
+		byteOrder.PutUint64(payload[:], holdTimeMs)
```
Another option here is to encode the payload as tlv for future extensibility.

joostjager replied:
Converted to tlv

joostjager commented on routing/payment_lifecycle.go:
```diff
@@ -554,11 +553,15 @@ func (p *shardHandler) collectResult(attempt *channeldb.HTLCAttemptInfo) (
+	resolutionFormat := attempt.Route.Hops[0].ResolutionFormat
+
+	errorDecryptor := htlcswitch.NewSphinxErrorDecrypter(
+		circuit, resolutionFormat,
```
One case to consider here is that if for some reason one of the nodes had to return `InvalidOnionPayload`, that node wasn't able to read the resolution format. Instead if will return an old-style error that we're not decoding properly. Not sure if we care because it should only happen in case of a bug?

joostjager commented on htlcswitch/hop/error_encryptor.go:
```diff
@@ -172,6 +213,30 @@ func (s *SphinxErrorEncrypter) Type() EncrypterType {
+	fatErrorBytes := [2]byte{
+		byte(s.fatErrors.MaxHops),
+		byte(s.fatErrors.MaxPayloadSize),
+	}
```
Here we can also decide to use tlv records.

joostjager replied:
Converted to tlv db records.

halseth commented on htlcswitch/hop/error_encryptor.go:
```diff
@@ -94,16 +102,21 @@ type SphinxErrorEncrypter struct {
+	}
+
+	if fatErrors != nil {
+		encrypter.CreatedAt = time.Now()
```
can set this field regardless of fat errors active or not

halseth replied:
or it has to do with the serialization logic that we don't want to set it in that case?

joostjager replied:
Yes the serialization logic, but indirectly. It wouldn't be a problem to set it, because serialization is gated on `s.fatErrors` which is still `nil`. Unit tests and their `DeepEqual` calls will fail on the creation time mismatch. Tried to make the comparison more sophisticated, but didn't turn out that nice. So decided to just not set `CreatedAt` here.

halseth reviewed (commented):
ACK - this turned out surprisingly clean! (I guess most of the complex logic is in lightning-onion) 👍

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

carlaKC commented on htlcswitch/hop/error_encryptor.go:
```diff
@@ -27,9 +27,9 @@ const (
-type ErrorEncrypterExtracter func(*btcec.PublicKey) (ErrorEncrypter,
+// SharedSecretGenerator defines a function signature that extracts a shared
+// secret from an sphinx OnionPacket.
+type SharedSecretGenerator func(*btcec.PublicKey) (sphinx.Hash256,
```
Could you give me the tl;dr for why we're changing this? Getting lost in the various layers of interfaces.

joostjager replied:
The problem is that without this refactor commit, `OnionProcessor` would need to know what kind of sphinx error encrypter to create in https://github.com/lightningnetwork/lnd/blob/17d7e84b26e3242e7f8447c8a40f0a12ce17400b/htlcswitch/hop/iterator.go#L403, where the scope at that point is really just the onion and the shared secret.

The circuit map calls into this method at https://github.com/lightningnetwork/lnd/blob/17d7e84b26e3242e7f8447c8a40f0a12ce17400b/htlcswitch/circuit_map.go#L636 after a restart. It might be necessary to get a attributable error encrypter from `OnionProcessor` then.

I've spent a crazy amount of time getting lost in the interfaces too, and also prototyped multiple approaches to finally settle on what I've got currently.

Will update the commit message with this reasoning.

carlaKC reviewed (commented):
approach ack!

joostjager commented on htlcswitch/hop/error_encryptor.go:
```diff
@@ -80,9 +95,11 @@ type ErrorEncrypter interface {
 // encryption and must be treated as such accordingly.
 type SphinxErrorEncrypter struct {
-	*sphinx.OnionErrorEncrypter
+	encrypter interface{}
```
Is it useful to define a marker interface in golang other than for documenting this specific field? I think it doesn't actually provide any compiler safety because every type will satisfy the interface?

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@joostjager, remember to re-request review from reviewers when ready

joostjager commented:
!lightninglabs-deploy mute

positiveblue commented on htlcswitch/hop/error_encryptor.go:
```diff
@@ -66,35 +89,126 @@ type ErrorEncrypter interface {
+	if errorStruct != nil {
+		// Set creation time rounded to nanosecond to avoid differences
+		// after serialization.
+		encrypter.CreatedAt = time.Now().Truncate(time.Nanosecond)
```
nit: I think `time.Now()` already [returns the a value in nanoseconds](https://pkg.go.dev/time#Time) (which happens to be the max granularity too)

Also, given that we will return the [hold time in ms](https://github.com/lightning/bolts/pull/1044/files#diff-ad65f0beaac5cef88f5fd7a8b9ca36cbc5a790f4815b4b50b9ea794a55aaf012R1052) and some databases cannot have a max granularity for timestamps of microseconds (I am thinking about [postgres](https://www.postgresql.org/docs/12/datatype-datetime.html)) should we truncate to micro/miliseconds here?

joostjager replied:
There was a reason why I had to do this. I think it was to strip off the monotonic clock reading.

joostjager commented:
Ran through a few of the single- and multi-hop inter-op scenarios with LDK. Seems to work.

saubyk commented:
Concept Ack

joostjager commented on htlcswitch/switch.go:
```diff
@@ -1322,16 +1322,24 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
+				fail.Reason, err = circuit.ErrorEncrypter.
+					IntermediateEncrypt(fail.Reason)
+				if err != nil {
+					err = fmt.Errorf("unable to obfuscate "+
```
Todo: create a proper failure message to attribute blame to the downstream peer. See https://github.com/lightningnetwork/lightning-onion/pull/60#discussion_r1228035385

joostjager replied:
Added correct handling when an intermediate node receives an invalid failure message from downstream.

joostjager commented on routing/pathfind.go:
```diff
@@ -125,6 +129,25 @@ func newRoute(sourceVertex route.Vertex,
 	)
 
+	// Determine the resolution format that is supported by the whole path.
+	attributableErrors := true
```
Todo: make this opt-in.

joostjager replied:
Added `--routerrpc.attr-errors`

joostjager commented:
Removed parameterization according to https://github.com/lightningnetwork/lightning-onion/pull/60#issuecomment-1593495056 and switched to 4 byte hmacs (down from 32 bytes). Also removed tlv payloads in favor of a fixed 32 bit hold time.

bitromortac reviewed (commented):
Great work! I think it's nice to have everything hard-coded to not reveal any fingerprint vector (except for the signal). Also, the upgrade path looks very reasonable to me, only when a majority of the network has upgraded to attributable errors (and a full attributable error route is likely), we'll request that error type (although some senders may prefer those routers already earlier). This will give a quick transition from zero senders to many senders and therefore a larger anonymity set.
From a routing node's perspective, attributable errors may not be so nice if they know they are slow and could get penalized for it, but if the network should have fast payments this may be needed?

joostjager commented on htlcswitch/switch.go:
```diff
@@ -1322,16 +1322,24 @@ func (s *Switch) handlePacketForward(packet *htlcPacket) error {
+				if err != nil {
+					err = fmt.Errorf("unable to obfuscate "+
+						"malformed error: %v", err)
+					log.Error(err)
```
What extra formatting would you like to see added here? The error itself is already a full sentence. I followed the approach that was already taken in master:

https://github.com/lightningnetwork/lnd/blob/f1bca2a59c6117cd2a23895d6b9c9a339904468b/htlcswitch/switch.go#L1311

joostjager commented on htlcswitch/hop/iterator.go:
```diff
@@ -383,9 +383,7 @@ func (p *OnionProcessor) DecodeHopIterators(id []byte,
-		p.router, ephemeralKey,
-	)
+	sharedSecret, err := p.router.GenerateSharedSecret(ephemeralKey)
 	if err != nil {
```
It is pre-existing. Maybe only `ErrInvalidOnionKey` is ever returned here. But not sure if we want to change this in this PR?

joostjager commented on htlcswitch/hop/error_encryptor.go:
```diff
@@ -159,7 +221,24 @@ func (s *SphinxErrorEncrypter) EncryptMalformedError(
+	switch {
+	// If the structure of the error received from downstream is invalid,
+	// then generate a new failure message with a valid structure so that
+	// the sender is able to penalize the offending node.
```
If they pass the short error, the upstream node will replace it by a proper error and the sender will indeed penalize this node.

joostjager commented on htlcswitch/hop/error_encryptor.go:
```diff
@@ -159,7 +221,24 @@ func (s *SphinxErrorEncrypter) EncryptMalformedError(
+		// Use an all-zeroes failure message. This is not a defined
+		// message, but the sender will at least know where the error
+		// occurred.
+		reason = make([]byte, lnwire.FailureMessageLength+2+2)
```
We could, although in the rest of the lnd repo we generally calculate that number 260 from `lnwire.FailureMessageLength` and I aligned this new code with that.

Semantically we'll indeed communicate zero message with zero padding. But it really does not matter what we put here.

As a future step, we could define a new failure code in BOLT 04 for this, but it won't change much about the penalization.

joostjager commented on htlcswitch/hop/iterator.go:
```diff
@@ -104,7 +104,8 @@ func (r *sphinxHopIterator) HopPayload() (*Payload, error) {
 // along with a failure code to signal if the decoding was successful. The
 // ErrorEncrypter is used to encrypt errors back to the sender in the event that
-// a payment fails.
+// a payment fails. This method always returns a legacy encrypter (no fat
```
Indeed. Fixed in the previous commit.

joostjager commented on channeldb/payments.go:
```diff
@@ -1297,6 +1303,13 @@ func deserializeHop(r io.Reader) (*route.Hop, error) {
 
+	attributableErrorType := uint64(record.AttributableErrorOnionType)
+	if _, ok := tlvMap[attributableErrorType]; ok {
+		delete(tlvMap, attributableErrorType)
```
Because right below here, the remaining records are stored as custom records. If we wouldn't delete here, the attr error record would also show up as a custom record.

joostjager commented:
>From a routing node's perspective, attributable errors may not be so nice if they know they are slow and could get penalized for it, but if the network should have fast payments this may be needed?

I think that ultimately they won't have a choice. At some point, I expect senders to avoid routing nodes that do not support attr errors.

joostjager commented:
Added integration test.

joostjager commented:
Updated `lightning-onion` dependency after adding updated test vector.

joostjager commented:
Updated feature bit to 36/37

joostjager commented on htlcswitch/link.go:
```diff
@@ -3029,9 +3029,11 @@ func (l *channelLink) processRemoteAdds(fwdPkg *channeldb.FwdPkg,
-		// encryption parameters.
+		// encryption parameters. Don't assume attributable errors,
+		// because first the resolution format needs to be decoded from
+		// the onion payload.
```
Fallback to legacy format is probably fine. If these happens to be an attributable error route, the upstream node will substitute with an attributable error?

thomash-acinq replied:
This should be added to the spec.
